### PR TITLE
OAuth integration

### PR DESCRIPTION
**Pull request by matthieusieben**
Running locally:
1) Run `pnpm i && pnpm build && pnpm dev` from `atproto`'s [feat-oauth-client](https://github.com/bluesky-social/atproto/tree/feat-oauth-client) branch
2) Copy the "🗼 Ozone service DID" value printed in the outpur
2) From this branch: `NEXT_PUBLIC_PLC_DIRECTORY_URL="http://localhost:2582" NEXT_PUBLIC_SOCIAL_APP_URL="http://localhost:2584" NEXT_PUBLIC_OZONE_SERVICE_DID="<ENTER_COPIED_OZONE_DID_HERE>" y dev`
4) Visit `http://127.0.0.1:3000` (make sure to use `127.0.0.1` and **not** `localhost`). Alternatively, start a reverse proxy using: `ngrok http 3000` and visit the proxy's URL instead

### CODE REVIEW

**Review by foysalit — COMMENTED** [review]
Leaving some comments after reviewing nearly 60% of the changes. Will do another pass once you get the rebased changes in.
Again, apologies for the huge conflict you're handling right now and LMK if I can help!

**Review by foysalit — COMMENTED** [review]
wow this is a massive piece of work! thank you for powering through the conflict resolution and it looks quite good. left some comments and questions but nothing major popped out. 

would love to get an extra pair of eyes from @devinivy and then do some testing with our mod team before creating a distro with this so let's hold on to merging until we get a distro out this week. will come back and approve once that's done.

File: app/actions/ModActionPanel/QuickAction.tsx
Comment (matthieusieben) [reply app/actions/ModActionPanel/QuickAction.tsx:168]: Yeah you need to work with the `feat-oauth-client` branch from the `atproto` repo. Requires some tweaks to work locally. I can explain how I do it if you need.

File: components/shell/ConfigurationContext.tsx
Comment (matthieusieben) [reply components/shell/ConfigurationContext.tsx:198]:
```ts 
const canFoo = usePermission('foo')
if (canFoo) {
  // OK
}
```

This is short hand for
```ts
const {serverConfig} = useConfigurationContext()
if (serverConfig.permissions["foo"]) {
  // OK
}
```

Both are exactly identical.

File: components/config/server-config.tsx
Comment (matthieusieben) [reply components/config/server-config.tsx:36]: `serverConfig` was completely split out from session management indeed.

If you look [here](https://github.com/bluesky-social/ozone/blob/oauth-matthieu/app/layout.tsx#L52-L67) you'll see that the `AuthProvider` wraps the `ConfigurationProvider`, and [here](https://github.com/bluesky-social/ozone/blob/oauth-matthieu/components/shell/ConfigurationContext.tsx#L88-L107) will show that the `serverConfig` will actually be loaded after auth was successful (`labelerAgent` available), and does not need to have been stored in the session at all.

File: app/layout.tsx
Comment (matthieusieben) [reply app/layout.tsx:53]: That's because the package was not released on NPM, meaning that types do not resolve properly in CI. Types are good locally ;-)

File: app/repositories/[id]/[...record]/page-content.tsx
Comment (matthieusieben) [reply app/repositories/[id]/[...record]/page-content.tsx:109]: You tell me. I kept the current behavior (that did not include the proxy headers):

https://github.com/bluesky-social/ozone/blob/134524d857a507600c002c6b14a8b51fbe06cc42/app/repositories/%5Bid%5D/%5B...record%5D/page-content.tsx#L106-L108

File: components/common/Tabs.tsx
Comment (matthieusieben) [reply components/common/Tabs.tsx:85]: I added prop to allow manipulating the current view from outside, allowing the parent component to sync the tab with the query params (if it wants to).

I did not "bake" this behavior here though a I believe this component should not concern itself with routing stuff for proper separation of concerns. wdyt?

File: components/list/Lists.tsx
Comment (matthieusieben) [reply components/list/Lists.tsx:24]: wrong ;-)
Capture d’écran 2024-07-09 à 11 10 05

Comment (matthieusieben) [reply components/list/Lists.tsx:24]: https://tanstack.com/query/latest/docs/framework/react/reference/useQuery#usequery

File: components/shell/AuthContext.tsx
Comment (matthieusieben) [reply components/shell/AuthContext.tsx:4]: Actually not deployed on NPM yet. This only works locally until `feat-oauth-client` branch from `atproto` repo is merged & published.

**Review by foysalit — APPROVED** [review]
Great work!

**Comment by matthieusieben** [comment]
Hey @foysalit, thanks for fixing this branch !

Note that a small [refactor](https://github.com/bluesky-social/atproto/pull/2714) due to a design change we made will break this branch once landed. The fix is trivial though.

File: middleware.ts
Comment (devinivy) [inline middleware.ts:17]: Why do we no longer need this middleware?

File: lib/constants.ts
Comment (devinivy) [inline lib/constants.ts:29]: The endpoint on the appview is going to be more reliable for this purpose.  Perhaps we can eventually implement this endpoint on the Ozone backend and it can serve itself.
```suggestion
export const HANDLE_RESOLVER_URL =
  process.env.NEXT_PUBLIC_HANDLE_RESOLVER_URL ||
  (process.env.NODE_ENV === 'development'
    ? 'http://localhost:2584'
    : 'https://api.bsky.app')
```

File: environment.d.ts
Comment (devinivy) [inline environment.d.ts:11]: I believe I saw an additional env related to handle resolution.

File: cypress/support/api.ts
Comment (devinivy) [inline cypress/support/api.ts:4]: I recommend using the appview for handle resolution:
```suggestion
export const HANDLE_RESOLVER_URL = 'https://api.bsky.app/xrpc'
```

File: components/shell/ConfigurationFlow.tsx
Comment (devinivy) [inline components/shell/ConfigurationFlow.tsx:56]: We specifically want the user to login using their service account, i.e. the account with the same DID as their labeler.  I don't think we need to iterate on the verbiage of the configuration flow as part of the OAuth integration, unless there's some new auth state that needs to be incorporated into the configuration flow.

Comment (devinivy) [reply components/shell/ConfigurationFlow.tsx:56]:
```suggestion
          {`You're`} not logged-in. Please login using your Ozone service
          account in order to configure Ozone.
```

File: app/oauth-client.json/route.ts
Comment (devinivy) [inline app/oauth-client.json/route.ts:25]:
```suggestion
      response_types: ['code'],
      grant_types: ['authorization_code', 'refresh_token'],
      token_endpoint_auth_method: 'none',
      scope: 'atproto transition:generic',
```

File: components/shell/CommandPalette/useAsyncSearch.tsx
Comment (devinivy) [inline components/shell/CommandPalette/useAsyncSearch.tsx:312]: Is this related to the OAuth integration, does it fix a bug, or a misc change/improvement?

File: components/shell/auth/oauth/useOAuth.ts
Comment (devinivy) [inline components/shell/auth/oauth/useOAuth.ts:132]: This gets a bit complicated to handle the branching paths of loading client metadata vs. baking it in.  I assume we just use one of these paths in ozone— can we remove the other?

File: lib/client-config.ts
Comment (devinivy) [inline lib/client-config.ts:29]:
```suggestion
    throw new Error('Mismatch between Ozone service DID and Ozone service metadata')
```

File: middleware.ts
Comment (devinivy) [reply middleware.ts:17]: Ah, found the client-side redirect 👍

**Review by devinivy — APPROVED** [review]

File: components/shell/auth/oauth/useOAuth.ts
Comment (matthieusieben) [reply components/shell/auth/oauth/useOAuth.ts]: Actually did that without even seeing your comment here 😅

File: middleware.ts
Comment (matthieusieben) [reply middleware.ts:17]: The redirect uri is `/`, which is conflicting with the logic here.

Also, the oauth flow uses the `state` param to "remember" the page that was being visited (in case credentials are no longer valid). This mechanism also conflicts with this middleware.

File: components/shell/CommandPalette/useAsyncSearch.tsx
Comment (matthieusieben) [reply components/shell/CommandPalette/useAsyncSearch.tsx:312]: @foysalit , I believe this was you ?

Comment (foysalit) [reply components/shell/CommandPalette/useAsyncSearch.tsx:312]: yep, this is a minor bugfix that I noticed while testing this branch so just put it in here. 
essentially, at-uris can contain handle instead of DID so this just makes sure that we resolve the handle to DID before doing the reroute since ozone always wants to use DIDs in URL.

File: lib/constants.ts
Comment (foysalit) [reply lib/constants.ts:11]: to add more context, it makes total sense that when developing locally, we'd want to point to local atproto dev env but often, I point at staging or production to check UI things which is adds extra step when committing changes etc. so, this is more of a devex consideration.

**Review by foysalit — APPROVED** [review]

File: lib/constants.ts
Comment (matthieusieben) [reply lib/constants.ts:11]: I added this so that it works with `yarn dev` without any additional `.env` file whatsoever. I am happy to remove this if that breaks your flow.

File: lib/client-config.ts
Comment (matthieusieben) [reply lib/client-config.ts:12]: This is solely for the purpose of making this work with `yarn dev`

**Pull request merged**
